[MLIR][ArmSVE] Add initial lowering of vector.contract to SVE `*MMLA` instructions

momchil-velikov opened the pull request:
Supersedes https://github.com/llvm/llvm-project/pull/135359

momchil-velikov committed:
[MLIR][ArmSVE] Add initial lowering of `vector.contract` to SVE `*MMLA` instructions

banach-space reviewed (commented):
Thanks! This one is a bit longer, so I may need to wait till Thursday before I can review.

One high-level question - would sharing some code between NEON and SVE be possible?

momchil-velikov commented:
> One high-level question - would sharing some code between NEON and SVE be possible?

No, I can't see it happening and resulting in less, or simpler, or easier to maintain code.
However, it might be possible to add Neon lowering to this patch and  see if the result is any good.

banach-space commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,304 @@
+// Check if the given value is a result of the operation `T` (which must be
+// sign- or zero- extend) from i8 to i32. Return the value before the extension.
+template <typename T>
+inline std::enable_if_t<(std::is_base_of_v<arith::ExtSIOp, T> ||
+                         std::is_base_of_v<arith::ExtUIOp, T>),
+                        std::optional<Value>>
```
Why not simple `isa<arith::ExtSIOp, arith::ExtUIOp>(v.getDefinitionOp())` inside the function instead of this? That's more common from what I've seen (there's very little SFINAE in the Dialect code).

momchil-velikov replied:
Because it does not need to be checked at runtime. But fair enough, it should be a `static_assert`, as there's no ambiguity to resolve, and such use is quite common across the code base, e.g.
```
  static_assert(llvm::is_one_of<OpTy, PackOp, UnPackOp>::value,
                "applies to only pack or unpack operations");
```

banach-space commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,304 @@
+  }
+}
+
+class LowerContractionToSVEI8MMPattern
```
It's a very long pattern. Could you document the high-level logic?

momchil-velikov replied:
Done.

banach-space commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,304 @@
+    mlir::VectorType rhsType = op.getRhsType();
+
+    // For now handle LHS<Mx8> and RHS<8x[N]> - these are the types we
+    // eventually expect from MMT4D. M and N dimensions must be even and at
```
[nit] We shouldn't be concerned with MMT4D in this dialect - it's a much higher-level abstraction and this logic should be valid irrespective of how the input is generated.

momchil-velikov replied:
There's no dependency on MMT4D - this comment is merely a rationale why we have chosen these particular operand shapes.

banach-space replied:
Perhaps just expand this comment a bit (e.g. by noting that MMT4D is the main use-case ATM)?

momchil-velikov replied:
Done (in the top-level description).

banach-space commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,304 @@
+    // least 2.
+    if (!lhsType.hasRank() || lhsType.getRank() != 2 || !rhsType.hasRank() ||
+        rhsType.getRank() != 2)
+      return failure();
```
Could you use `notifyMatchFailure` with some descriptive error message instead? Thanks! Some comment for other instances of `failure`.

banach-space commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,304 @@
+    // For now handle LHS<Mx8> and RHS<8x[N]> - these are the types we
+    // eventually expect from MMT4D. M and N dimensions must be even and at
+    // least 2.
+    if (!lhsType.hasRank() || lhsType.getRank() != 2 || !rhsType.hasRank() ||
```
IIRC, inputs to `vector.contract` are required to be vectors, hence `lhsType.hasRank()` should always be true, no?

momchil-velikov replied:
Done

banach-space commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,304 @@
+      return failure();
+
+    // One-dimensional vector types for arm_sve.*mmla
+    auto nxv16i8 = VectorType::get(16, rewriter.getI8Type(), {true});
```
Similar suggestion for other instances of `VectorType::get`
```suggestion
    auto nxv16i8 = VectorType::get(/*shape=*/16, rewriter.getI8Type(), /*scalableDims=*/{true});
```

momchil-velikov replied:
Sure, I changed it, but IMHO it's exceedingly repetitive, overkill and makes the source harder to read.

banach-space commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,304 @@
+          loc, r0, r1,
+          llvm::ArrayRef<int64_t>{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,
+                                  14, 15});
+      // Turn it into a scalable vector.
```
[nit] This is "broadcasting", so a bit more involved/nuanced than just turning into a scalable vector. Perhaps expand?

momchil-velikov replied:
No, the "broadcasting" is the following operation, this indeed only turns the fixed vector into a scalable one, possibly padded with poison.

banach-space commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,304 @@
+    auto nxv16i8 = VectorType::get(16, rewriter.getI8Type(), {true});
+    auto nxv4i32 = VectorType::get(4, rewriter.getI32Type(), {true});
+
+    // Extract LHS sub-tiles.
```
[nit] Could you specify the dims? That would be helpful.

momchil-velikov replied:
Done.

banach-space commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,304 @@
+// This file implements lowering patterns from vector.contract to
+// SVE I8MM operations.
+//
+//===---
```
```suggestion
//===----------------------------------------------------------------------===//```

banach-space commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,304 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements lowering patterns from vector.contract to
+// SVE I8MM operations.
```
Could you add a note that `vector.contract` needs to be accompanied by `arith.extsi` (or `arith.extui`) Ops? Also, is I8MM the official name? Shouldn't that be FEAT_I8MM? 

Basically,  could we document a bit more?

banach-space reviewed (commented):
Thanks Momchil - this is great!

I skimmed through the pattern logic, and it's very neatly written. It's actually quite easy to follow, despite the underlying logic being a bit convoluted - well done! I've left a few minor suggestions, but nothing major.

Also, it seems like we should be able to extend this fairly easily to support NEON as well. Worth thinking about 🙂

Now, overall this patch is quite large, and I’d suggest extracting the end-to-end / integration tests into a separate PR. Additionally, the remaining tests currently use `--convert-vector-to-llvm=`, which lowers all the way to LLVM (i.e., it exercises a lot of patterns). Instead, I’d recommend testing `LowerContractionToSVEI8MMPattern` in isolation and only verifying that the correct sequence of ArmSVE ops (plus some Vector ops) is generated - for example:

```mlir
    (...)
    %33 = arm_sve.smmla %23, %7, %15 : vector<[16]xi8> to vector<[4]xi32>
    %34 = arm_sve.smmla %24, %7, %16 : vector<[16]xi8> to vector<[4]xi32>
    %35 = arm_sve.smmla %31, %13, %15 : vector<[16]xi8> to vector<[4]xi32>
    %36 = arm_sve.smmla %32, %13, %16 : vector<[16]xi8> to vector<[4]xi32>
```

That way, we will:
* reduce noise in the test output (by focusing on a single pattern),
* simplify expected output (fewer ops to match),
* avoid re-testing functionality already covered elsewhere (e.g., `arm_sve.smmla` → `arm_sve.intr.smmla` lowering).

Btw, this is already looking great, and I know I’m asking for a bit of a rewrite (especially around the tests), but I really think it’ll help with long-term maintainability.

dcaballe reviewed (commented):
Really excited to see this! I'll take a look in the next iteration. Thanks!

momchil-velikov committed:
[fixup] Misc changes

  -- come commenting
  -- replace enable_if with a staic assert
  -- return reasons for match failures

momchil-velikov committed:
[fixup] Handle implicit sign-extend of LHS and RHS

momchil-velikov committed:
[fixup] Detach integration tests

They'll come in a separate PR.

banach-space commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,358 @@
+    mlir::VectorType lhsType = op.getLhsType();
+    mlir::VectorType rhsType = op.getRhsType();
+
+    // Check the operands have the expected shape. M and N dimensions must be
+    // even and at least 2.
```
> and at least 2

Note that this is checking that the rank is 2 rather than any of the dimensions being at least "2"

momchil-velikov replied:
The comment applies for code further down.

banach-space replied:
Lets just move it closer to the code that it documents.

banach-space commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,358 @@
+        N % 2 != 0 || !rhsType.getScalableDims()[0])
+      return rewriter.notifyMatchFailure(op, "non-matching operand shape");
+
+    // Check permutation maps. For now only accept
```
[nit] Add a note that this is a regular matmul with RHS transposed.

banach-space commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,358 @@
+    }
+
+    // "Flatten" the RHS tile from <[N]x8> to <[8*N]>.
+    auto RHS = rewriter.create<vector::ShapeCastOp>(
```
I like `RHS`, but we should stick with the naming convention and use `camelBack`. Same for `M`, `N` and `K` :(

momchil-velikov replied:
I don't mind `lhs`,`rhs`,` etc.

IMHO `M`, `N` and `K` are fine the way they are.

They have a wide scope thus not very suitable for a  single letter lowercase names.

Longer names (e.g. `lhsRows`) are imprecise as all of them are rows and cols in more than one matrix (`lhsRows`  == `outRows` == `accRows`, `lhsCols` == `rhsCols` (or maybe `rhsRows` depending on whether one refers to logical or memory layout dimensions), `rhsCols` == `outCols` == `accCols` (or maybe it's `rhsRows`). Just creates confusion.

banach-space replied:
> IMHO M, N and K are fine the way they are.

I am a bit uneasy about diverging from the coding guidelines, but in this case it would hurt readability, lets keep it as is. This is complex as is.

banach-space commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,358 @@
+        outTile.push_back(mmla);
+      }
+
+    // Unpack the OUT sub-tiles and insert into the result.
```
Would you mind adding some ASCII here?

banach-space commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,358 @@
+    auto accRowX264Ty = VectorType::get(/*shape=*/N, rewriter.getI64Type(),
+                                        /*scalableDims=*/{true});
+
+    // Extract and pack the ACC sub-tiles.
```
ASCII pls :)

banach-space reviewed (commented):
A few more questions/suggestions inline. In general, I would appreciate a few more comments with ASCII - these tend to be super helpful :) 🙏🏻 

One other high-level comment, why not incorporate the test changes from https://github.com/llvm/llvm-project/pull/140572 in this patch? If 140572 implements the end state (that would be my preference) then lets use that. The TD op can still be upstreamed as a separate change.

All in all, this is already very polished and nicely crafted, well done!

momchil-velikov commented on mlir/lib/Dialect/ArmNeon/Transforms/LowerContractionToSMMLAPattern.cpp:
```diff
@@ -56,6 +56,9 @@ class LowerContractionToSMMLAPattern
     // Avoid 0-D vectors and 1-D rhs:
     if (!lhsType.hasRank() || !rhsType.hasRank() || rhsType.getRank() < 2)
       return failure();
+    // Avoid scalable vectors.
```
Eh? This rewrite will not work with scalable vectors regardless of whether there's another rewrite or not.

Should be a separate PR, in principle, but it's hardly worth the faff.

banach-space replied:
Let me rephrase.

The comment “Avoid scalable vectors.” explains **what** the code is doing, which is already clear from the code itself. What would be more helpful here is a brief note on **why** scalable vectors are being avoided. This may be obvious to those familiar with NEON and SVE, but adding a bit of context would make the intent clearer for others as well.

Relatedly, it might be worth mentioning that there's a separate implementation for NEON and SVE. Even within Arm, not everyone is aware that I8MM is available for both.

All I am asking for is a small addition to the comment.

momchil-velikov replied:
Done.

momchil-velikov commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,358 @@
+  return inOp;
+}
+
+// Designate the operation (resp. instruction) used to do sub-tile matrix
+// multiplications.
```
I don't think so. There's a choice between `arm_sve::SmmlaOp`, `arm_sve::UmmlaOp`, and `arm_sve::UsmmlaOp` which are separate operations and not variants of anything (resp. for instructions).

banach-space replied:
To me, these are just different variants of MMLA. However, after your comment, I see that others might read this differently.

momchil-velikov commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,358 @@
+//   * if the operand is of i8 type and the operation is sign-extend, return the
+//     operand itself.
+//
+// This way we handle both explicit sign- or zero- extension or implicit
+// sign-extension.
```
https://github.com/llvm/llvm-project/blob/51e222ef48222e6e5745b96a484164dbdcaea239/mlir/include/mlir/Dialect/Vector/IR/VectorOps.td#L58

momchil-velikov commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,358 @@
+// for the left-hand side and the right-hand side, respectively,
+// yielding a <Mx[N]> 32-bit integer result.
+//
+// The operands shapes are such that the operands can be evenly split into
```
Done.

momchil-velikov committed:
[fixup] Commenting and some minor refactoring

kiranchandramohan commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,367 @@
+    // Extract LHS sub-tiles with logicall shape <2x8>.
+    SmallVector<Value> lhsTile;
+    for (int64_t i = 0; i < M; i += 2) {
+      // Exract two consective rows of the LHS tile.
```
```suggestion
      //Extract two consecutive rows of the LHS tile.
```

kiranchandramohan reviewed (commented):
A few Nit comments.

momchil-velikov commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,367 @@
+  MixedSwapped // usmmla with LHS and RHS swapped
+};
+
+// Create the matrix multply and accumulate operation according to `op`.
```
Done.

momchil-velikov commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,367 @@
+
+        // Interleave the rows, effectively flattening each 2x2 tile into 4
+        // consecutive elements.
+        auto intr_i64 =
```
Done.

momchil-velikov commented on mlir/lib/Dialect/ArmSVE/Transforms/LowerContractionToSVEI8MMPattern.cpp:
```diff
@@ -0,0 +1,367 @@
+Value createMMLA(PatternRewriter &rewriter, MMLA op, Location loc,
+                 mlir::VectorType accType, Value acc, Value lhs, Value rhs) {
+  switch (op) {
+  case MMLA::Signed:
+    return rewriter.create<arm_sve::SmmlaOp>(loc, accType, acc, lhs, rhs);
+  case MMLA::Unsigned:
+    return rewriter.create<arm_sve::UmmlaOp>(loc, accType, acc, lhs, rhs);
+  case MMLA::Mixed:
+    return rewriter.create<arm_sve::UsmmlaOp>(loc, accType, acc, lhs, rhs);
+  case MMLA::MixedSwapped:
+    // The accumulator comes transposed and the result will be transposed
+    // later, so all we have to do here is swap the operands.
+    return rewriter.create<arm_sve::UsmmlaOp>(loc, accType, acc, rhs, lhs);
```
The using-directive not needed, removed.

momchil-velikov committed:
[fixup] Minor style/spelling fixes

dcaballe reviewed (approved):
Thanks!
